src/login/auth_session.cpp

```diff
@@ -88,41 +94,124 @@ void auth_session::do_read()
+        }
+    };
 
-    std::memcpy(usernameBuffer, buffer_.data() + 0x09, 16);
-    std::memcpy(passwordBuffer, buffer_.data() + 0x19, 32);
-    // 1 byte of command at 0x39
-    const std::string version = asStringFromUntrustedSource(buffer_.data() + 0x61, 5);
+    auto sendJsonOnlyErrorCode = [&](const std::string& errorMessage)
+    {
+        json loginErrorCodeReply;
+        loginErrorCodeReply["result"]        = -1; // xiloader requires there to be a value here on json parsing
+        loginErrorCodeReply["error_message"] = errorMessage;
 
-    std::string username{ usernameBuffer };
-    std::string password{ passwordBuffer };
+        sendJsonAsBuffer(loginErrorCodeReply);
+    };
 
-    // Only match on the first 3 characters of the version string
-    // ie. 1.1.1 -> 1.1.x
-    // Major.Minor.Patch
-    // Major and minor version changes should be breaking, patch should not.
-    if (strncmp(version.c_str(), SUPPORTED_XILOADER_VERSION, 3) != 0)
+    if (jsonBuffer.is_discarded()) // not json
     {
-        ref<uint8>(buffer_.data(), 0) = LOGIN_ERROR_VERSION_UNSUPPORTED;
+        const auto newModeFlag = ref<uint8>(buffer_.data(), 0) == 0xFF;
+        if (!newModeFlag)
+        {
+            ShowDebug("Old xiloader connected. Not supported.");
+            ref<uint8>(buffer_.data(), 0) = LOGIN_ERROR;
+            do_write(1);
+            return;
+        }
 
-        do_write(1);
-        return;
+        // Feature flags from xiloader are sent over on the 2nd byte+
+
+        char usernameBuffer[17] = {};
+        char passwordBuffer[33] = {};
+
+        std::memcpy(usernameBuffer, buffer_.data() + 0x09, 16);
+        std::memcpy(passwordBuffer, buffer_.data() + 0x19, 32);
+
+        // 1 byte of command at 0x39
+        const std::string versionStr = asStringFromUntrustedSource(buffer_.data() + 0x61, 5);
+        updated_password             = asStringFromUntrustedSource(buffer_.data() + 0x40, 32);
+
+        username = usernameBuffer;
+        password = passwordBuffer;
+
+        // Only match on the first 3 characters of the version string
+        // ie. 1.1.1 -> 1.1.x
+        // Major.Minor.Patch
+        // Major and minor version changes should be breaking, patch should not.
+        if (strncmp(versionStr.c_str(), SUPPORTED_XILOADER_VERSION, 3) != 0)
+        {
+            ref<uint8>(buffer_.data(), 0) = LOGIN_ERROR_VERSION_UNSUPPORTED;
+
+            do_write(1);
+            return;
+        }
+        code = ref<uint8>(buffer_.data(), 0x39);
     }
+    else
+    {
+        isJson = true;
+
+        auto maybeUsername        = loginHelpers::getJsonStringOrDefault(jsonBuffer, "username");
+        auto maybePassword        = loginHelpers::getJsonStringOrDefault(jsonBuffer, "password");
+        auto maybeUpdatedPassword = loginHelpers::getJsonStringOrDefault(jsonBuffer, "new_password");
+        auto maybeCode            = loginHelpers::getJsonTrivialTypeOrDefault<decltype(code)>(jsonBuffer, "command");
```

Comment:
So these aren't "OrDefault" anymore if they're using `std::optional`. Do you need to be using `decltype(code)`, when `code` is defined as an `int8` at the start of this function? I would change all of these to look like:
```cpp
loginHelpers::jsonGet<T>(buffer, key) -> std::optional<T>;
```

So you can use them as: `loginHelpers::jsonGet<std::string>(buffer, key) -> std::optional<std::string>;`

Then inside of `jsonGet` you don't need to use `enable_if`, you can use `auto` and then `constexpr if` like I have in `database.h` for a much easier read with a lot less duplication. Ping me on discord if you want to talk about this some more

Reply:
Yeah, these used to actually be "OrDefault" for real but then I switched to std::optional so completely valid here